code/modules/mob/living/carbon/xenomorph/abilities.dm

```diff
@@ -439,6 +439,176 @@
+
+/datum/action/ability/activable/xeno/secrete_special_resin/give_action(mob/living/L)
+	. = ..()
+	var/mutable_appearance/build_maptext = mutable_appearance(icon = null,icon_state = null, layer = ACTION_LAYER_MAPTEXT)
+	build_maptext.pixel_x = 0
+	build_maptext.pixel_y = 25
+	build_maptext.maptext = MAPTEXT(GLOB.hive_datums[owner.get_xeno_hivenumber()].special_build_points)
+	visual_references[VREF_MUTABLE_SPECIAL_RESIN_COUNTER] = build_maptext
+
+/datum/action/ability/activable/xeno/secrete_special_resin/update_button_icon()
+	var/mob/living/carbon/xenomorph/X = owner
+	var/atom/A = X.selected_special_resin
+	action_icon_state = initial(A.name)
+	button.cut_overlay(visual_references[VREF_MUTABLE_SPECIAL_RESIN_COUNTER])
+	var/mutable_appearance/number = visual_references[VREF_MUTABLE_SPECIAL_RESIN_COUNTER]
+	number.maptext = MAPTEXT("[GLOB.hive_datums[owner.get_xeno_hivenumber()].special_build_points]")
+	visual_references[VREF_MUTABLE_SPECIAL_RESIN_COUNTER] = number
+	button.add_overlay(visual_references[VREF_MUTABLE_SPECIAL_RESIN_COUNTER])
+	return ..()
+
+/datum/action/ability/activable/xeno/secrete_special_resin/action_activate()
+	//Left click on the secrete resin button opens up radial menu (new type of changing structures).
+	var/mob/living/carbon/xenomorph/X = owner
+	if(X.selected_ability != src)
+		return ..()
+	. = ..()
+	var/resin_choice = show_radial_menu(owner, owner, GLOB.resin_special_images_list, radius = 35)
+	if(!resin_choice)
+		return
+	var/i = GLOB.resin_images_list.Find(resin_choice)
+	X.selected_special_resin = buildable_special_structures[i]
+	var/atom/A = X.selected_special_resin
+	X.balloon_alert(X, initial(A.name))
+	update_button_icon()
+
+/datum/action/ability/activable/xeno/secrete_special_resin/alternate_action_activate()
+	//Right click on secrete resin button cycles through to the next construction type (old method of changing structures).
+	var/mob/living/carbon/xenomorph/X = owner
+	if(X.selected_ability != src)
+		return ..()
+	var/i = buildable_special_structures.Find(X.selected_special_resin)
+	if(length(buildable_special_structures) == i)
+		X.selected_special_resin = buildable_special_structures[1]
+	else
+		X.selected_special_resin = buildable_special_structures[i+1]
+	var/atom/A = X.selected_special_resin
+	X.balloon_alert(X, initial(A.name))
+	update_button_icon()
+
+/datum/action/ability/activable/xeno/secrete_special_resin/use_ability(atom/A)
+	var/mob/living/carbon/xenomorph/xowner = owner
+	if(get_dist(owner, A) > xowner.xeno_caste.resin_max_range) //Maximum range is defined in the castedatum with resin_max_range, defaults to 0
+		build_special_resin(get_turf(owner))
+	else
+		build_special_resin(get_turf(A))
+
+/datum/action/ability/activable/xeno/secrete_special_resin/proc/get_wait()
+	. = base_wait
+	if(!scaling_wait)
+		return
+	var/mob/living/carbon/xenomorph/X = owner
+
+	var/build_resin_modifier = 1
+	switch(X.selected_resin)
+		if(/obj/alien/resin/sticky)
+			build_resin_modifier = 0.5
+
```

Comment:
This is not checking the special resin var (aka, if you have sticky selected on your base action, your special walls will build twice as fast).